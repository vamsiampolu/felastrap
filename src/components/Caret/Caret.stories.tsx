import { Caret } from "./Caret";
import type { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof Caret> = {
  component: Caret,
  title: "components/Caret",
  tags: ["autodocs"],
  parameters: {
    layout: ["centered"],
  },
};

export default meta;

export type Story = StoryObj<typeof Caret>;

export const Down = {
  args: {
    width: "0.3em",
    direction: "down",
  },
};

export const Up = {
  args: {
    width: "0.3em",
    direction: "up",
  },
};

export const End = {
  args: {
    width: "0.3rem",
    direction: "end",
  },
};

export const Start = {
  args: {
    width: "0.3rem",
    direction: "start",
  },
};
