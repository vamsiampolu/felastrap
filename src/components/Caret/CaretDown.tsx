import { useFela } from "react-fela";
import type { Theme, StyleProps, CaretProps } from "../common";

const caret = ({ width }: StyleProps<CaretProps>) => ({
  borderBottom: 0,
  borderLeft: `${width} solid transparent`,
  borderRight: `${width} solid transparent`,
  borderTop: `${width} solid`,
  content: "",
  display: "inline-block",
  marginLeft: `calc(${width} * 0.85)`,
  verticalAlign: `calc(${width} * 0.85)`,
});

export const CaretDown = ({ width = "0.3em", ...other }: CaretProps) => {
  const { css } = useFela<Theme, CaretProps>({ width });
  return <div className={css(caret)} {...other}></div>;
};
