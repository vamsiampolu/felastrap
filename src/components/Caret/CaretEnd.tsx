import { useFela } from "react-fela";
import type { Theme, StyleProps, CaretProps } from "../common";

const caret = ({ width }: StyleProps<CaretProps>) => ({
  borderTop: `${width} solid transparent`,
  borderRight: 0,
  borderBottom: `${width} solid transparent`,
  borderLeft: `${width} solid`,
  content: "",
  display: "inline-block",
  marginLeft: `calc(${width} * 0.85)`,
  verticalAlign: `calc(${width} * 0.85)`,
});

export const CaretEnd = ({ width = "0.3em", ...other }: CaretProps) => {
  const { css } = useFela<Theme, CaretProps>({ width });
  return <div className={css(caret)} {...other}></div>;
};
