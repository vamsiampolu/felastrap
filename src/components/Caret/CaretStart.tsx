import { useFela } from "react-fela";
import type { Theme, StyleProps, CaretProps } from "../common";

const caret = ({ width }: StyleProps<CaretProps>) => ({
  display: "inline-block",
  marginLeft: `calc(${width} * 0.85)`,
  marginRight: "3px",
  marginBottom: "-2px",
  verticalAlign: `calc(${width} * 0.85)`,
  content: "",
  borderTop: `${width} solid transparent`,
  borderRight: `${width} solid`,
  borderBottom: `${width} solid transparent`,
});

export const CaretStart = ({ width = "0.3em", ...other }: CaretProps) => {
  const { css } = useFela<Theme, CaretProps>({ width });
  return <div className={css(caret)} {...other}></div>;
};
