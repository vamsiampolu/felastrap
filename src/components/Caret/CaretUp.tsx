import { useFela } from "react-fela";
import type { Theme, StyleProps, CaretProps } from "../common";

const caret = ({ width }: StyleProps<CaretProps>) => ({
  borderTop: 0,
  borderRight: `${width} solid transparent`,
  borderBottom: `${width} solid`,
  borderLeft: `${width} solid transparent`,
  content: "",
  display: "inline-block",
  marginLeft: `calc(${width} * 0.85)`,
  verticalAlign: `calc(${width} * 0.85)`,
});

export const CaretUp = ({ width = "0.3em", ...other }: CaretProps) => {
  const { css } = useFela<Theme, CaretProps>({ width });
  return <div className={css(caret)} {...other}></div>;
};
