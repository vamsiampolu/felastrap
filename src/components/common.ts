import type * as Css from "csstype";
import type * as Fela from "fela";
import type {
  ComponentPropsWithoutRef,
  ComponentType,
  ElementType,
} from "react";
import type * as ReactFela from "react-fela";

export type ExcludeFrom<T, U> = T extends U ? never : T;
export type Theme = object;

export type Rule = Fela.TRule;
export type CSSObject = Fela.CSSObject;
export type StyleProps<P = object> = ReactFela.StyleProps<Theme, P>;
export type StyleFunction<P = object> = ReactFela.StyleFunction<Theme, P>;

export type Padding = Required<
  Pick<
    Css.Properties,
    "paddingTop" | "paddingLeft" | "paddingRight" | "paddingBottom"
  >
>;
export type BorderRadius = Required<
  Pick<
    Css.Properties,
    | "borderTopLeftRadius"
    | "borderTopRightRadius"
    | "borderBottomRightRadius"
    | "borderBottomLeftRadius"
  >
>;

export type PaddingString = Required<Css.Properties["padding"]>;
export type Color = Css.Property.Color;

export type HeaderTypes = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
export type Size = "small" | "medium" | "large";
export type Variant =
  | "default"
  | "danger"
  | "info"
  | "primary"
  | "secondary"
  | "success"
  | "warning";

export type FontSize = Css.Property.FontSize;
export type ColorMapping = Record<Variant, Color>;
export type PaddingBySize = Record<Size, Padding>;

export type SpanProps = ComponentPropsWithoutRef<"span">;
export type DivProps = ComponentPropsWithoutRef<"div">;
export type ListItemProps = ComponentPropsWithoutRef<"li">;

export type DropdownDirection = "up" | "down" | "end" | "start" | "centered";
export type CaretDirection = "up" | "down" | "end" | "start";
export type NavVariant =
  | "default"
  | "tabs"
  | "pills"
  | "underline"
  | "vertical";

export type CaretProps = DivProps & {
  width: string | number | null;
};

export type DynamicElement<T extends ElementType = "span"> =
  | keyof JSX.IntrinsicElements
  | ComponentType<T>;
export type PargraphProps = ComponentPropsWithoutRef<"p">;
export type ImageProps = ComponentPropsWithoutRef<"img">;
export type TooltipPlacement = "top" | "left" | "bottom" | "right";

export type MediaQuerySize = "xs" | "sm" | "md" | "lg" | "xl" | "xxl";
export type Timer = ReturnType<typeof setTimeout>;
export type OnClick<T extends ElementType> =
  ComponentPropsWithoutRef<T>["onClick"];
export type StyleKey = Exclude<
  keyof HTMLElement["style"],
  "length" | "parentRule"
>;
