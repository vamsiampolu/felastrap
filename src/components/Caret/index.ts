/* istanbul ignore file */
import { Caret, type CaretWrapperProps as CaretProps } from "./Caret";

export { Caret, type CaretProps };
