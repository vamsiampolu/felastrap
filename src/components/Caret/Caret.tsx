import { CaretDown } from "./CaretDown";
import { CaretUp } from "./CaretUp";
import { CaretEnd } from "./CaretEnd";
import { CaretStart } from "./CaretStart";
import type { CaretDirection, DivProps } from "../common";

export type CaretWrapperProps = DivProps & {
  direction?: CaretDirection;
  width?: string;
};

export const Caret = ({
  direction = "down",
  width = "0.3em",
  ...other
}: CaretWrapperProps) => {
  switch (direction) {
    case "up":
      return <CaretUp width={width} {...other} />;

    case "end":
      return <CaretEnd width={width} {...other} />;

    case "start":
      return <CaretStart width={width} {...other} />;

    default:
      return <CaretDown width={width} {...other} />;
  }
};
