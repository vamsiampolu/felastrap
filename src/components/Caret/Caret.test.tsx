import { IRenderer } from "fela";
import { Caret } from "./Caret";
import { createRenderer, cssReboot } from "../../test-helpers/";
import { CaretDirection } from "../common";
import { createSnapshot } from "../../test-helpers/createSnapshot";

describe("components/Caret/Caret", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render a caret", async () => {
    const component = <Caret />;
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it.each(["down", "end", "start", "up"])(
    "should render a caret with direction %s",
    async (direction) => {
      const component = (
        <Caret direction={direction as CaretDirection} width="0.3rem" />
      );
      const snapshot = await createSnapshot({ component, renderer });
      expect(snapshot).toMatchSnapshot();
    },
  );

  it.each(["down", "end", "start", "up"])(
    "should render a caret with direction %s and default width",
    async (direction) => {
      const component = <Caret direction={direction as CaretDirection} />;
      const snapshot = await createSnapshot({ component, renderer });
      expect(snapshot).toMatchSnapshot();
    },
  );

  it.each(["down", "end", "start", "up"])(
    "should render a caret with direction %s and custom width",
    async (direction) => {
      const component = (
        <Caret direction={direction as CaretDirection} width="0.5rem" />
      );
      const snapshot = await createSnapshot({ component, renderer });
      expect(snapshot).toMatchSnapshot();
    },
  );
});
