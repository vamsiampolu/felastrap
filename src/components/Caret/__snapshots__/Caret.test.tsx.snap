// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Caret/Caret should render a caret 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.3em * 0.85);
}
.d {
  vertical-align: calc(0.3em * 0.85);
}
.e {
  border-top-width: 0.3em;
}
.f {
  border-top-style: solid;
}
.g {
  border-right-width: 0.3em;
}
.h {
  border-right-style: solid;
}
.i {
  border-right-color: transparent;
}
.j {
  border-left-width: 0.3em;
}
.k {
  border-left-style: solid;
}
.l {
  border-left-color: transparent;
}
.m {
  border-bottom-width: 0;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction down 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.3rem * 0.85);
}
.d {
  vertical-align: calc(0.3rem * 0.85);
}
.e {
  border-top-width: 0.3rem;
}
.f {
  border-top-style: solid;
}
.g {
  border-right-width: 0.3rem;
}
.h {
  border-right-style: solid;
}
.i {
  border-right-color: transparent;
}
.j {
  border-left-width: 0.3rem;
}
.k {
  border-left-style: solid;
}
.l {
  border-left-color: transparent;
}
.m {
  border-bottom-width: 0;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction down and custom width 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.5rem * 0.85);
}
.d {
  vertical-align: calc(0.5rem * 0.85);
}
.e {
  border-top-width: 0.5rem;
}
.f {
  border-top-style: solid;
}
.g {
  border-right-width: 0.5rem;
}
.h {
  border-right-style: solid;
}
.i {
  border-right-color: transparent;
}
.j {
  border-left-width: 0.5rem;
}
.k {
  border-left-style: solid;
}
.l {
  border-left-color: transparent;
}
.m {
  border-bottom-width: 0;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction down and default width 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.3em * 0.85);
}
.d {
  vertical-align: calc(0.3em * 0.85);
}
.e {
  border-top-width: 0.3em;
}
.f {
  border-top-style: solid;
}
.g {
  border-right-width: 0.3em;
}
.h {
  border-right-style: solid;
}
.i {
  border-right-color: transparent;
}
.j {
  border-left-width: 0.3em;
}
.k {
  border-left-style: solid;
}
.l {
  border-left-color: transparent;
}
.m {
  border-bottom-width: 0;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction end 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.3rem * 0.85);
}
.d {
  vertical-align: calc(0.3rem * 0.85);
}
.e {
  border-left-width: 0.3rem;
}
.f {
  border-left-style: solid;
}
.g {
  border-bottom-width: 0.3rem;
}
.h {
  border-bottom-style: solid;
}
.i {
  border-bottom-color: transparent;
}
.j {
  border-right-width: 0;
}
.k {
  border-top-width: 0.3rem;
}
.l {
  border-top-style: solid;
}
.m {
  border-top-color: transparent;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction end and custom width 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.5rem * 0.85);
}
.d {
  vertical-align: calc(0.5rem * 0.85);
}
.e {
  border-left-width: 0.5rem;
}
.f {
  border-left-style: solid;
}
.g {
  border-bottom-width: 0.5rem;
}
.h {
  border-bottom-style: solid;
}
.i {
  border-bottom-color: transparent;
}
.j {
  border-right-width: 0;
}
.k {
  border-top-width: 0.5rem;
}
.l {
  border-top-style: solid;
}
.m {
  border-top-color: transparent;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction end and default width 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.3em * 0.85);
}
.d {
  vertical-align: calc(0.3em * 0.85);
}
.e {
  border-left-width: 0.3em;
}
.f {
  border-left-style: solid;
}
.g {
  border-bottom-width: 0.3em;
}
.h {
  border-bottom-style: solid;
}
.i {
  border-bottom-color: transparent;
}
.j {
  border-right-width: 0;
}
.k {
  border-top-width: 0.3em;
}
.l {
  border-top-style: solid;
}
.m {
  border-top-color: transparent;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction start 1`] = `
".a {
  display: inline-block;
}
.b {
  margin-left: calc(0.3rem * 0.85);
}
.c {
  margin-right: 3px;
}
.d {
  margin-bottom: -2px;
}
.e {
  vertical-align: calc(0.3rem * 0.85);
}
.f {
  content:;
}
.g {
  border-bottom-width: 0.3rem;
}
.h {
  border-bottom-style: solid;
}
.i {
  border-bottom-color: transparent;
}
.j {
  border-right-width: 0.3rem;
}
.k {
  border-right-style: solid;
}
.l {
  border-top-width: 0.3rem;
}
.m {
  border-top-style: solid;
}
.n {
  border-top-color: transparent;
}


<div class="a b c d e f g h i j k l m n"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction start and custom width 1`] = `
".a {
  display: inline-block;
}
.b {
  margin-left: calc(0.5rem * 0.85);
}
.c {
  margin-right: 3px;
}
.d {
  margin-bottom: -2px;
}
.e {
  vertical-align: calc(0.5rem * 0.85);
}
.f {
  content:;
}
.g {
  border-bottom-width: 0.5rem;
}
.h {
  border-bottom-style: solid;
}
.i {
  border-bottom-color: transparent;
}
.j {
  border-right-width: 0.5rem;
}
.k {
  border-right-style: solid;
}
.l {
  border-top-width: 0.5rem;
}
.m {
  border-top-style: solid;
}
.n {
  border-top-color: transparent;
}


<div class="a b c d e f g h i j k l m n"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction start and default width 1`] = `
".a {
  display: inline-block;
}
.b {
  margin-left: calc(0.3em * 0.85);
}
.c {
  margin-right: 3px;
}
.d {
  margin-bottom: -2px;
}
.e {
  vertical-align: calc(0.3em * 0.85);
}
.f {
  content:;
}
.g {
  border-bottom-width: 0.3em;
}
.h {
  border-bottom-style: solid;
}
.i {
  border-bottom-color: transparent;
}
.j {
  border-right-width: 0.3em;
}
.k {
  border-right-style: solid;
}
.l {
  border-top-width: 0.3em;
}
.m {
  border-top-style: solid;
}
.n {
  border-top-color: transparent;
}


<div class="a b c d e f g h i j k l m n"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction up 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.3rem * 0.85);
}
.d {
  vertical-align: calc(0.3rem * 0.85);
}
.e {
  border-left-width: 0.3rem;
}
.f {
  border-left-style: solid;
}
.g {
  border-left-color: transparent;
}
.h {
  border-bottom-width: 0.3rem;
}
.i {
  border-bottom-style: solid;
}
.j {
  border-right-width: 0.3rem;
}
.k {
  border-right-style: solid;
}
.l {
  border-right-color: transparent;
}
.m {
  border-top-width: 0;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction up and custom width 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.5rem * 0.85);
}
.d {
  vertical-align: calc(0.5rem * 0.85);
}
.e {
  border-left-width: 0.5rem;
}
.f {
  border-left-style: solid;
}
.g {
  border-left-color: transparent;
}
.h {
  border-bottom-width: 0.5rem;
}
.i {
  border-bottom-style: solid;
}
.j {
  border-right-width: 0.5rem;
}
.k {
  border-right-style: solid;
}
.l {
  border-right-color: transparent;
}
.m {
  border-top-width: 0;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;

exports[`components/Caret/Caret should render a caret with direction up and default width 1`] = `
".a {
  content:;
}
.b {
  display: inline-block;
}
.c {
  margin-left: calc(0.3em * 0.85);
}
.d {
  vertical-align: calc(0.3em * 0.85);
}
.e {
  border-left-width: 0.3em;
}
.f {
  border-left-style: solid;
}
.g {
  border-left-color: transparent;
}
.h {
  border-bottom-width: 0.3em;
}
.i {
  border-bottom-style: solid;
}
.j {
  border-right-width: 0.3em;
}
.k {
  border-right-style: solid;
}
.l {
  border-right-color: transparent;
}
.m {
  border-top-width: 0;
}


<div class="a b c d e f g h i j k l m"></div>
;
"
`;
